### ISSUE TITLE
Make core transform isomorphic to run in plugin context also

### ISSUE THREAD

**Issue by kwonoj** [swc-project/swc#4826]
### Describe the feature

There are some possible needs for the plugin who want to visit some state of AST _after_ a certain transform phase completes. 

In the case of custom transform (via before_custom_pass) it is possible to run any existing core transform visitor as needed prior to run custom pass, then run swc core's transform. It is probably not efficient in some aspects, but it is allowed. However, some visitors may not be allowed to chain certain visitor in the plugin context. Notably some rely on globals, handlers, and other few more plugin sdk provides proxy instead of host. 

Ideally, we'd like to provide identical experiences between plugin / custom pass. We may need to update few core transform visitors, mostly under _compat to use proper context to allow this. Probably this is ad-hoc bases as someone need to use it, instead of trying to make everything at once.

In case of babel since context is always same under node.js, plugin author can freely duck-typing like 

```
//plugin

const firstPass = transform(code, { plugins: [someBabelCorePlugins]);

// actual visitor
firstPass.visit()
```

In our case it's not possible for the plugin.

**Comment by kdy1** [comment]
I think it's better to allow using `swc_common::errors::HANDLER` and related friends from plugin, rather than using proxy api from core transforms

**Comment by overlookmotel** [comment]
Is this intended to support the case where e.g. I have a custom plugin which I want to run *after* transform of JSX to JS?

If that were possible, it would simplify some custom plugins a lot.

**Comment by kwonoj** [comment]
> Is this intended to support the case where e.g. I have a custom plugin which I want to run after transform of JSX to JS?

No. This'll allow some core transforms run in plugins instead. The order of transform between core / plugin won't change regardless.

**Comment by overlookmotel** [comment]
Ah I see. Thanks for clarifying. But is it the intent that a custom plugin can call an SWC transform pass itself (e.g. JSX to JS) before running it's own custom visitor? This is what I'd understood from "it is possible to run any existing core transform visitor as needed prior to run custom pass" - but maybe I'm misunderstanding?

**Comment by kwonoj** [comment]
> custom plugin can call an SWC transform pass itself (e.g. JSX to JS) before running it's own custom visitor?

Yes, that's correct. But again, this doesn't mean anything aronud actual transform execution order between core / plugin and if core transform omits based on plugin's behavior or not. Certain plugin can try some core transform in its transform pass, but the execution order of plugin / core transform itself won't change.

**Comment by cyjake** [comment]
any updates regarding this issue? just stumbled upon the same problem, wish there is a way to make the plugin transform take place after core transforms finish. Or calling core transform directly at the plugin transform phase in advance.

**Comment by kdy1** [comment]
You can invoke core transforms from the plugin transform

**Comment by cyjake** [comment]
I tried to process the program directly with react() pass, but got stuck with the `Lrc<SourceMap>` dependency. It seems in the plugin context, only `PluginSourceMapProxy` can be used, which currently is readonly and cannot be used to access vanilla SourceMap object. Just throwing an empty SourceMap shouldn't be an option, right?

**Comment by kdy1** [comment]
@cyjake Can you send a PR to change the type?

**Comment by cyjake** [comment]
@kdy1 I did a shallow dig into the swc code base, here are my two cents. There are two argument that swc_ecma_transform_react::react() requires but not available yet in the plugin executor context, 1) an `Lrc<SourceMap>` that is both read- and write-able, 2) the top_level_mark. 

Regarding the first one, it looks like that to get the needed SourceMap, I'd have to implement the SourceMap trait by declaring functions that are actually used in `react()` transform. Namely, new_source_file(), look_up_char_pos(), and span_to_snippet(). Some of them aren't implemented in the PluginSourceMapProxy yet.

Currently I've got little experience implementing these api, feel free to point any necessary references if you think it still is worth a shot.
